src/kafka-extensions/registry.ts

```diff
@@ -0,0 +1,185 @@
+import * as vscode from "vscode";
+import { KafkaConfig } from "kafkajs";
+import { Cluster, ConnectionOptions, createDefaultKafkaConfig as createDefaultKafkaConfig } from "../client/client";
+import { ClusterSettings } from "../settings/clusters";
+import { configureDefaultClusters } from "../wizards/clusters";
+import { ClusterProviderParticipant, KafkaExtensionParticipant } from "./api";
+
+/**
+ * Cluster provider is used to:
+ *
+ *  - collect clusters (eg: create a cluster from a wizard, import clusters from a repository, ...)
+ * and add them to the Kafka Explorer.
+ *  - configure a Kafka client from a complex process (eg : use SSO to connect to the cluster)
+ *
+ * Implementing a cluster provider in a 3rd party extension is done in 2 steps:
+ *
+ *  - define the cluster provider (id, name) in the package.json in the contributes/kafka/clusterProviders section.
+ *  - return the cluster provider participant (See KafkaExtensionParticipant) to use in the activate() of the extension.
+ *
+ */
+export class ClusterProvider {
+
+    private processor: ClusterProviderParticipant | undefined;
+
+    constructor(private definition: ClusterProviderDefinition, private extensionId: string) {
+
+    }
+
+    /**
+     * Returns the cluster provider id.
+     */
+    public get id(): string {
+        return this.definition.id;
+    }
+
+    /**
+     * Returns the cluster provider name.
+     */
+    public get name(): string {
+        return this.definition.name || this.definition.id;
+    }
+
+    /**
+     * Returns the clusters managed by the provider which must be added to the kafka explorer.
+     *
+     * @param clusterSettings the cluster settings.
+     */
+    async collectClusters(clusterSettings: ClusterSettings): Promise<Cluster[] | undefined> {
+        const processor = await this.getProcessor();
+        return processor.configureClusters(clusterSettings);
+    }
+
+    /**
+     * Create the Kafka JS client instance from the given connection options.
+     *
+     * @param connectionOptions the connection options.
+     */
+    async createKafkaConfig(connectionOptions: ConnectionOptions): Promise<KafkaConfig | undefined> {
+        const processor = await this.getProcessor();
+        if (processor.createKafkaConfig) {
+            return processor.createKafkaConfig(connectionOptions);
+        }
+    }
+
+    private async getProcessor(): Promise<ClusterProviderParticipant> {
+        if (this.processor) {
+            return this.processor;
+        }
+        // The cluster provider processor is not already loaded, try to activate the owner extension.
+        // The return of the extension activate() method must return the cluster provider processor.

```

Comment:
The extension's activate() method must return the kafka extension participant.

Reply:
fixed